### PR TITLE
Add M3 goal planner pipeline

### PR DESCRIPTION
**Pull request by spyroot**
## Summary

- Add the M3 Goal Planner / Goal Extractor that maps operator text plus context into structured goal-plan JSON.
- Add a real-corpus M3 JSONL dataset and collator for causal-LM SFT.
- Add a redfish_ctl capture builder that reads the existing `~/.json_responses` contract, recursively merges `rest_api_map.npy` method maps, filters non-Redfish manifest JSON, and emits derived action/goal-plan records without raw response bodies.
- Add an M3 trainer and CLI for `infer`, `train`, `export-redfish-actions`, and `build-redfish-ctl-dataset`.

## Validation

- `ruff check igc/modules/m3_goal_planner.py igc/ds/m3_goal_plan_dataset.py igc/ds/m3_redfish_goal_plan_builder.py igc/modules/m3_goal_planner_train.py igc/modules/m3_goal_planner_cli.py` -> `All checks passed!`
- `python -m compileall -q ...` on the five new M3 files -> passed.
- Real local capture build: `build-redfish-ctl-dataset --json-root /Users/spyroot/.json_responses --output-jsonl /private/tmp/m3_goal_plans_real.jsonl --templates-per-action 2` -> `records=5224`, `actions=2612`.
- CPU smoke trainer: `gpt2`, 32 real records, 3 optimizer steps -> saved `/private/tmp/m3_goal_planner_real_smoke`, final logged loss `1.7869453430175781`.
- One-row real-data overfit: `gpt2`, 40 optimizer steps -> loss dropped from `2.2177236080169678` to `0.014077377505600452`.
- No-fallback model inference from the overfit checkpoint parsed generated M3 JSON and matched the training target path/action for the real Redfish row.
- High-level text interface check for `boot server with ubuntu2204 and adjust bios for fast boot` produced a structured plan with `mount_virtual_media`, `set_boot_override`, `set_bios_attribute`, pending-settings apply, reboot-if-required, and final verification steps.

Generated JSONL and checkpoint artifacts were kept under `/private/tmp` and are not committed.

### CODE REVIEW

**Comment by spyroot** [comment]
Heads-up from the integration pass: main now has #100, which fixes the long-dead CI gate (it ran bare `pytest` → `No module named 'igc'` on every module; now `python -m pytest` + `pynvml` in test deps + lazy deepspeed). **Rebase this branch on current `main`** and the offline collection that was failing here should pass.

One thing to verify while rebasing: after #100, any offline test that loads a real HuggingFace model via `from_pretrained` must be marked `@pytest.mark.download` (otherwise it fails in a clean CI env with `LocalEntryNotFoundError`, passing locally only via a warm cache). Please confirm `tests/gpu/test_m3_goal_planner_distributed_logging.py` and any other m3 test that instantiates a backbone is marked appropriately (`gpu`/`download`/`slow` per the taxonomy in `pytest.ini`). This PR is large (+2949/16 files) so it's queued for human review rather than an unattended merge.

**Comment by spyroot** [comment]
Closing per the binding M3/GoalExtractor correction (TEAM_GUIDE, 2026-07-15): not the active source-of-truth implementation; must not merge as a separate planner island. Its engineering lessons were ported into the shared Phase 1/2/3 pipeline (#110-#113). Branch preserved; archive-tagged archive/codex-m3-goal-planner.
